### PR TITLE
#1304 Add thrown exceptions to the generated nested mapping methods

### LINKED ISSUE #1304: Exceptions are not being declared to be thrown
At the request of @filiphr, I have created a barebones project to reproduce the issue I was having with the generated mapping files not declaring exceptions to be thrown.

Please refer to the following repository/project: https://github.com/xyzst/unwrapped-exceptions

### PR DESCRIPTION
**Pull request by filiphr**
Fixes #1304.

Exceptions that can be thrown in assignments in nested methods are added to the throws clause to the generated nested methods as well

### CODE REVIEW

File: processor/src/test/java/org/mapstruct/ap/test/nestedbeans/exceptions/EntityFactory.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+            entity = entityClass.newInstance();
+        }
+        catch ( IllegalAccessException exception ) {
+            throw new MappingException( "Rare exception thrown, refer to stack trace", exception );
+        }
+        catch ( InstantiationException exception ) {
+            throw new MappingException( "Rare exception thrown, refer to stack trace", exception );
+        }
+        catch ( Exception exception ) {
+            throw new MappingException( "I don't know how you got here, refer to stack trace", exception );
```
Comment (xyzst) [inline processor/src/test/java/org/mapstruct/ap/test/nestedbeans/exceptions/EntityFactory.java]: 😂

Love how you kept the original message

Comment (filiphr) [reply processor/src/test/java/org/mapstruct/ap/test/nestedbeans/exceptions/EntityFactory.java]: Yes :). I'll add your name to the authors as well :). I didn't take everything in order to reduce the amount of code that we bring in

File: processor/src/test/java/org/mapstruct/ap/test/nestedbeans/exceptions/_target/DeveloperDto.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+    private String firstName;
+
+    private String lastName;
+
+    private String department;
+
+    private Integer experience;
+
+    private Double salary;
```
Comment (sjaakd) [inline processor/src/test/java/org/mapstruct/ap/test/nestedbeans/exceptions/_target/DeveloperDto.java]: `Double` or `BigDecimal`? :smile:

File: processor/src/test/java/org/mapstruct/ap/test/nestedbeans/exceptions/_target/DeveloperDto.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package org.mapstruct.ap.test.nestedbeans.exceptions._target;
+
+/**
+ * @author Filip Hrisafov
+ * @author Darren Rambaud
+ */
+public class DeveloperDto {
```
Comment (sjaakd) [inline processor/src/test/java/org/mapstruct/ap/test/nestedbeans/exceptions/_target/DeveloperDto.java]: Are all these properties needed to demonstrate the problem (both in parent and child)? They confused me initially..

File: processor/src/test/java/org/mapstruct/ap/test/nestedbeans/exceptions/_target/DeveloperDto.java (hunk 2 above)
Comment (filiphr) [reply processor/src/test/java/org/mapstruct/ap/test/nestedbeans/exceptions/_target/DeveloperDto.java]: Living on the edge 😄. I see you've been surrounded by these lately :)

File: processor/src/test/java/org/mapstruct/ap/test/nestedbeans/exceptions/_target/DeveloperDto.java (hunk 3 above)
Comment (filiphr) [reply processor/src/test/java/org/mapstruct/ap/test/nestedbeans/exceptions/_target/DeveloperDto.java]: You are right. Not everything is needed. I'll remove all except one

**Comment by gunnarmorling** [comment]
Seems we could cut down a bit those props as outlined by @sjaakd? Otherwise it LGTM. Go ahead and merge if you like. Thanks!

**Comment by sjaakd** [comment]
The first commit name seems a bit strange..

File: processor/src/test/java/org/mapstruct/ap/test/nestedbeans/exceptions/_target/DeveloperDto.java (hunk 3 above)
Comment (sjaakd) [reply processor/src/test/java/org/mapstruct/ap/test/nestedbeans/exceptions/_target/DeveloperDto.java]: Great. I think we can save the example for other cases (I'm always out of inspiration when I need test cases and hence end up with complete aquaria).

Comment (filiphr) [reply processor/src/test/java/org/mapstruct/ap/test/nestedbeans/exceptions/_target/DeveloperDto.java]: Actually @xyzst made a more extensive example [here](https://github.com/xyzst/unwrapped-exceptions). Maybe he can contribute it to our examples repo if he is up for it :)

**Comment by filiphr** [comment]
@sjaakd which commit message? The one with fixup? That is not the first one. The first one is "#1304 Add thrown exceptions to the generated nested mapping methods"

**Pull request merged**